### PR TITLE
[RPA-685.2] Set proper status of interrupted process instances & its' events.

### PR DESCRIPTION
**Pull request by star-in-the-jar**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-scheduler/src/database/process-instance-event/process-instance-event.service.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ export class ProcessInstanceEventService {
         return queryRunner.manager.findOne(ProcessInstanceEventEntity, { where: { executionId }, relations: ['processInstance'] });
     }
 
-    findActiveByProcessInstanceId(processInstanceId: string) {
+    async findActiveByProcessInstanceId(processInstanceId: string) {
```
Comment (FranekMarciniak) [inline runbotics/runbotics-scheduler/src/database/process-instance-event/process-instance-event.service.ts]: .

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 2]
```diff
@@ -1,20 +1,26 @@
...
 export class BotLifecycleService {
     constructor(
         private readonly processInstanceService: ProcessInstanceService,
         private readonly processInstanceEventService: ProcessInstanceEventService,
-        private readonly botProcessEventService: BotProcessEventService,
     ) { }
     
     private readonly logger = new Logger(BotLifecycleService.name);
 
+    private setInfoPropsAfterBotDisconnection(process) {
```
Comment (FranekMarciniak) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: type

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 3]
```diff
@@ -23,40 +29,72 @@ export class BotLifecycleService {
...
     }
 
-    private async handleInterruptedProcessInstanceEvent (bot: IBot, processInstanceId: string): Promise<void> {
-        const activeProcessInstanceEvents = this.processInstanceEventService.findActiveByProcessInstanceId(processInstanceId);
-        const newProcessInstanceEvent = {
-            ...activeProcessInstanceEvents,
-            status: ProcessInstanceEventStatus.ERRORED,
-            error: 'Bot has been shut down',
-        };
+    private async handleInterruptedProcessInstanceEvent (processInstanceId: string): Promise<void> {
```
Comment (FranekMarciniak) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: this should be a transaction

**Review by FranekMarciniak — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-scheduler/src/database/process-instance-event/process-instance-event.service.ts (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/database/process-instance-event/process-instance-event.service.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 2 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 3 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: done

**Review by star-in-the-jar — COMMENTED** [review]
Self reviewed

**Review by star-in-the-jar — COMMENTED** [review]
Self reviewed

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 4]
```diff
@@ -67,9 +69,14 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
     ) {
         const installationId = socket.bot.installationId;
+
         this.logger.log(`=> Updating process-instance (${processInstance.id}) by bot (${installationId}) | status: ${processInstance.status}`);
         await this.botProcessService.updateProcessInstance(installationId, processInstance);
         this.logger.log(`<= Success: process-instance (${processInstance.id}) updated by bot (${installationId}) | status: ${processInstance.status}`);
+
+        if(processInstance.status !== ProcessInstanceStatus.TERMINATED) return;
+
+        this.changeEventStatusOnInstanceTermination(socket.bot, processInstance);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: I'm not a fan of it. The last log was meant to be the last operation of this function to have a better understanding of app state when debugging prod/dev logs

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 5]
```diff
@@ -124,4 +123,34 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
+                    id: processInstance.id,
+                    process: {
+                        id: processInstance.process.id
+                    }
+                }
+            };
+
+            await this.updateProcessInstanceEvent(bot, newProcessInstanceEvent);
+        });
+    }
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: this method should be in one of the services. Treat the gateway class like a controller class, business logic should be moved to another layer such as a service

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 6]
```diff
@@ -124,4 +123,34 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
         await this.botProcessEventService.updateProcessInstanceEvent(processInstanceEvent, bot);
         this.logger.log(`<= Success: process-instance-event (${processInstanceEvent.executionId}) updated by bot (${installationId}) | step: ${processInstanceEvent.step}, status: ${processInstanceEvent.status}`);
     }
+
+    private async changeEventStatusOnInstanceTermination(bot: IBot, processInstance: IProcessInstance, retryCount = 0) {
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: avoid linking words like `on` in the names of functions or variables, there is a better way

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 7]
```diff
@@ -124,4 +123,34 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
+
+        const activeEvents: IProcessInstanceEvent[] = await this.processInstanceEventService.findActiveByProcessInstanceId(processInstance.id);
+        
+        if(activeEvents.length === 0) {
+            this.logger.log(`<= No active events of process-instance (${processInstance.id}) were found. Retrying (${retryCount + 1}/3)`);
+            setTimeout(() => this.changeEventStatusOnInstanceTermination(bot, processInstance, retryCount + 1), 1000);
+            return;
+        }
+
+        await activeEvents.forEach(async (event) => {
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: forEach returns a void, await is redundant

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 8]
```diff
@@ -124,4 +123,34 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
+    private async changeEventStatusOnInstanceTermination(bot: IBot, processInstance: IProcessInstance, retryCount = 0) {
+        if(retryCount === 3) {
+            this.logger.log(`<= No active events of process-instance (${processInstance.id}) were found. Aborting.`);
+            return;
+        }
+
+        const activeEvents: IProcessInstanceEvent[] = await this.processInstanceEventService.findActiveByProcessInstanceId(processInstance.id);
+        
+        if(activeEvents.length === 0) {
+            this.logger.log(`<= No active events of process-instance (${processInstance.id}) were found. Retrying (${retryCount + 1}/3)`);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: there will be just `1/3` for default `retryCount` value

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 9]
```diff
@@ -124,4 +123,34 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
+            setTimeout(() => this.changeEventStatusOnInstanceTermination(bot, processInstance, retryCount + 1), 1000);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: I dont like this idea of `setTimeout`. This function won't be predictable

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 10]
```diff
@@ -1,62 +1,110 @@
...
                 processInstance.status !== ProcessInstanceStatus.IN_PROGRESS
             ) return;
 
-            const newProcessInstance = { 
-                ...processInstance,
-                status: ProcessInstanceStatus.ERRORED,
-                error: 'Bot has been shut down',
+            const newProcessInstance: IProcessInstance = {
+                ...processInstance, 
+                status: ProcessInstanceStatus.ERRORED,
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts:47]: empty space at the end

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 11]
```diff
@@ -1,62 +1,110 @@
...
-    private async handleInterruptedProcessInstanceEvent (bot: IBot, processInstanceId: string): Promise<void> {
-        const activeProcessInstanceEvents = this.processInstanceEventService.findActiveByProcessInstanceId(processInstanceId);
-        const newProcessInstanceEvent = {
-            ...activeProcessInstanceEvents,
-            status: ProcessInstanceEventStatus.ERRORED,
-            error: 'Bot has been shut down',
-        };
+    private async handleProcessInstanceEventInterruption (processInstance: IProcessInstance, bot: IBot): Promise<void> {
+        const activeEvents: IProcessInstanceEvent[] = await this.processInstanceEventService.findActiveByProcessInstanceId(processInstance.id);
+        const processId: number = await this.processInstanceService.findById(processInstance.id).then(processInstance => processInstance.process.id);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: make a new line for each chain

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 12]
```diff
@@ -1,62 +1,110 @@
...
+        
+        if(!activeEvents) return;
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts:62]: this guard should be higher, redundant DB call for processId

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 13]
```diff
@@ -1,62 +1,110 @@
...
+
+        this.logger.log(
+            `=> Updating process-instance (${
+                processInstance.id
+            }) updated after bot disconnection | status: ${
+                processInstance.status
+            }, error message: "${
+                processInstance.error
+            }"`
+        );
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: there is a misconception. This kind of logs with arrows are meant to mean the incoming and outgoing info but here there are other operations

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 14]
```diff
@@ -1,62 +1,110 @@
...
+        
+        this.logger.log(
+            `<= Success process-instance-event (${
+                processInstanceEvent.id
+            }) has been updated after bot disconnection | status: ${
+                processInstanceEvent.status
+            }, error message: "${
+                processInstanceEvent.error
+            }"`
+        );
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: same here

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts (hunk 8 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: if it does find any active events (which is 95% odds) it does not show log at all. 

If it doesn't it shows this:

image

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts (hunk 7 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts (hunk 6 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 10 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts:47]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 12 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts:62]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 13 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: so are the "error message" or the "updated after bot disconnection" wrong?

bot.gateway uses updateProcessInstance too

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 14 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: done i guess

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 13 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: changed, take a look

**Review by star-in-the-jar — COMMENTED** [review]
self reviewed

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 15]
```diff
@@ -1,62 +1,80 @@
...
-                processInstance.status !== ProcessInstanceStatus.IN_PROGRESS
-            ) return;
-
-            const newProcessInstance = { 
-                ...processInstance,
-                status: ProcessInstanceStatus.ERRORED,
-                error: 'Bot has been shut down',
-            };
+    async handleProcessInstanceInterruption(bot: IBot): Promise<void> {
+        const disconnectedInstance: IProcessInstance = await this.processInstanceService.findActiveByBotId(bot.id);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: change to find all

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 16]
```diff
@@ -1,62 +1,80 @@
...
+    private async updateInterruptedProcessInstance(processInstance: IProcessInstance, installationId: string) {
+        this.logger.log(`=> Updating process-instance (${processInstance.id}) after bot disconnection | status: ${processInstance.status}`);
+        await this.botProcessService.updateProcessInstance(installationId, processInstance);
+        this.logger.log(`<= Success process-instance (${processInstance.id}) updated | status: ${processInstance.status}`);
+    }
+    
+    private async updateInterruptedProcessInstanceEvent(processInstanceEvent: IProcessInstanceEvent, bot: IBot) {
+        this.logger.log(`=> Updating process-instance-event (${processInstanceEvent.id}) | status: ${processInstanceEvent.status}`);
+        await this.botProcessEventService.updateProcessInstanceEvent(processInstanceEvent, bot);
+        this.logger.log(`<= Success process-instance-event (${processInstanceEvent.id}) updated | status: ${processInstanceEvent.status}`);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: remove arrows in log messages + add info about interruption

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 17]
```diff
@@ -67,9 +69,17 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
         this.logger.log(`=> Updating process-instance (${processInstance.id}) by bot (${installationId}) | status: ${processInstance.status}`);
         await this.botProcessService.updateProcessInstance(installationId, processInstance);
         this.logger.log(`<= Success: process-instance (${processInstance.id}) updated by bot (${installationId}) | status: ${processInstance.status}`);
+
+        if(
+            processInstance.status !== ProcessInstanceStatus.ERRORED &&
+            processInstance.status !== ProcessInstanceStatus.TERMINATED
+        ) return;
+
+        this.setEventStatusesAlikeInstance(socket.bot, processInstance);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: the outgoing log should be the last operation in controller handler

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts [hunk 18]
```diff
@@ -124,4 +126,26 @@ export class BotWebSocketGateway implements OnGatewayDisconnect, OnGatewayConnec
...
+                ...event, 
+                status: newStatus,
+                finished: processInstance.updated,
+                processInstance,
+            };
+
+            this.updateProcessInstanceEvent(bot, newProcessInstanceEvent);
+        });
+    }
+
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: move it to the related service class

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts [hunk 19]
```diff
@@ -60,21 +61,47 @@ export class BotProcessEventService {
...
+                await queryRunner.manager
+                .findOne(
+                    ProcessInstanceEntity,
+                    { where: { id: processInstanceEvent.processInstance.id } }
                 )
-                .execute();
-
-            await this.updateProcessInstance(queryRunner, processInstanceEvent);
+                .catch(error => {
+                    this.logger.log(`process-instance (${processInstanceEvent.processInstance.id}) not found | Error: ${error});`);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts]: log.error

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts [hunk 20]
```diff
@@ -22,9 +22,10 @@ export const getProcessInstanceUpdateFieldsByStatus = (
             if (dbStatus) return STARTING_PROCESS_INSTANCE_FIELDS;
             return ['status', ...STARTING_PROCESS_INSTANCE_FIELDS];
         }
-
         case ProcessInstanceStatus.COMPLETED:
         case ProcessInstanceStatus.ERRORED:
+            if(dbStatus === ProcessInstanceStatus.IN_PROGRESS || dbStatus === undefined) return COMPLETED_PROCESS_INSTANCE_FIELDS;
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts]: if there is no reason to make this change then revert incoming changes

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 15 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: done - still working despite many instances with status IN_PROGRESS (but does not fix the old ones)

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts (hunk 16 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts (hunk 17 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts (hunk 18 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/bot.gateway.ts]: done

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts (hunk 19 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts]: done

**Review by preludium — DISMISSED** [review]

**Review by preludium — APPROVED** [review]

**Pull request merged**